### DISCUSSION TITLE
Use of an I2C digital IO expander for UI buttons?

### DISCUSSION THREAD

**Discussion by diyelectromusic** [probonopd/MiniDexed#509] [Ideas]
Here is an initial branch of code that will allow the use of a cheap PCF8574 8-way I2C digital IO expander for UI buttons.

https://github.com/diyelectromusic/MiniDexed/commits/i2cbuttons

What do we think - is this useful?  It would mean that up to 8 buttons can be interfaced to the board without using up extra GPIO pins which might be in use for other things... or might just simplify the wiring in a build.

It's just an idea anyway.

As an aside, this is the same type of I2C IO expander that is used for the I2C version of the HD44780 display... and in the photo below you can see it working fine alongside a I2C SSD1306 display.

Kevin

![IMG_7219](https://github.com/probonopd/MiniDexed/assets/68612569/607b9eef-ac71-4866-91a6-7fc39e0e4f5b)

**Comment by u77345** [comment]
I'm adding Zynthian-like hardware support to Circle, to be used here. (MCP23017, ILI9341, WM8731). Would PCF8574 be a candidate for the same strategy for /addon/gpio, with a generalized GPIO interface? 
My work in progress is at https://github.com/u77345/circle/blob/master/addon/gpio/mcp23017.cpp

**Reply by diyelectromusic** [reply]
Potentially, but tbh it would be a significant over complication for the PCF8574 which has just a single 8-bit register to read over I2C...

But if the MCP23017 was available, I'd definitely have a look at that too.

Kevin

**Comment by probonopd** [comment]
While this is cool, I think it may complicate the code and parts count. Unless there is a killer use case, I'd say this shouldn't go into the main branch, which I am trying(!) to keep lean.

**Reply by diyelectromusic** [reply]
The use case is freeing up other GPIO, allowing more buttons, and simplifying building as it means buttons can be installed on a case more remotely with only four wires back to the Pi rather than one per button (and GND).

In terms of code, it is a minor extension to the code that already exists for button and MIDI button handling (see the diffs), so isn't a major addition to the code and certainly pretty trivial compared to some of the other changes already there around performances, effects, and the like, and other proposals re graphics and such.

But of course its up to you :)

Kevin

**Comment by diyelectromusic** [comment]
For what it's worth this branch now supports the PCF8575 16-bit expander too.

Let me know if this is of interest.

Kevin
